hasktorch/src/Torch/Typed/Functional.hs

```diff
@@ -3008,8 +3008,52 @@ mv input vec = unsafePerformIO $ ATen.cast2 ATen.Managed.mv_tt input vec
 -- mvlgamma :: Tensor device dtype shape -> Int -> Tensor device dtype shape
 -- mvlgamma _input _p = unsafePerformIO $ (ATen.cast2 ATen.Managed.mvlgamma_tl) _input _p
 
--- narrow :: Tensor device dtype shape -> Int -> Int -> Int -> Tensor device dtype shape
--- narrow _input _dim _start _length = unsafePerformIO $ (ATen.cast4 ATen.Managed.narrow_tlll) _input _dim _start _length
+type family
+  NarrowCheck
+    (mbCurrent :: Maybe Nat)
+    (mbUpdated :: Maybe [Nat])
+    (shape :: [Nat])
+    (dim :: Nat)
+    (start :: Nat)
+    (length :: Nat) ::
+    [Nat]
+  where
+  NarrowCheck Nothing _ sh d _ _        = DimOutOfBound sh d
+  NarrowCheck (Just c) Nothing sh d s l = DimOutOfBound sh d
+  NarrowCheck _ (Just r) _ _ _ _        = r
+
+type family Narrow (dim :: Nat) (shape :: [Nat]) (current :: Maybe Nat) (start :: Nat) (length :: Nat) :: Maybe [Nat] where
+  Narrow d sh (Just c) s l = ReplaceDim d sh ((c - s) + l)
+  Narrow d sh Nothing s l =
+    TypeError
+      ( Text "Requested narrow dimension "
+          :<>: ShowType d
+          :<>: Text " doesnt exist in "
+          :<>: ShowType sh
+      )
+
+type family NarrowInDomain (mbCurrent :: Maybe Nat) (start :: Nat) (length :: Nat) :: Constraint where
+  NarrowInDomain (Just c) s l = c >= (s + l)
+
+-- | "Narrow" a tensor by returning a tensor that is a slice from 'start' of length 'length' along 'dim'
+-- 
+-- >>> narrow @0 @0 @2 (ones :: CPUTensor 'D.Float '[3,3,3])
+-- Tensor Float [2,3,3] 
+-- >>> narrow @1 @1 @2 (ones :: CPUTensor 'D.Float '[3,3,3])
+-- Tensor Float [3,2,3] 
+narrow :: forall dim start length shape mbSize mbNewShape dtype device. 
+  (All KnownNat '[dim, start, length]
+  , All KnownNat shape
+  , DimOutOfBoundCheck shape dim
+  , mbSize ~ ExtractDim dim shape
+  , NarrowInDomain mbSize start length
+  , mbNewShape ~ Narrow dim shape mbSize start length) =>
```

Comment:
Nice work!
One suggestion I’d like to make is that you should look into refactoring these shape constraints into a single type or type family. Otherwise, a caller will have to list all of them.

Reply:
done